pkg/commands/container.go

```diff
@@ -341,22 +348,101 @@ func (c *Container) Restart() error {
 }
 
 // Attach attaches the container
-func (c *Container) Attach() (*exec.Cmd, error) {
+func (c *Container) Attach() error {
 	// verify that we can in fact attach to this container
 	if !c.Details.Config.OpenStdin {
-		return nil, errors.New(c.Tr.UnattachableContainerError)
+		return errors.New(c.Tr.UnattachableContainerError)
 	}
 
 	if c.Container.State == "exited" {
-		return nil, errors.New(c.Tr.CannotAttachStoppedContainerError)
+		return errors.New(c.Tr.CannotAttachStoppedContainerError)
 	}
 
-	cmd := c.OSCommand.PrepareSubProcess("docker", "attach", "--sig-proxy=false", c.ID)
-	return cmd, nil
+	options := types.ContainerAttachOptions{
+		Stream: true,
+		Stdin:  c.Details.Config.OpenStdin,
+		Stdout: true,
+		Stderr: true,
+		Logs:   true,
+	}
+
+	hijack, err := c.Client.ContainerAttach(context.Background(), c.ID, options)
+	if err != nil {
+		return err
+	}
+
+	fd := os.Stdin.Fd()
+	oldState, err := term.MakeRaw(fd)
+	if err != nil {
+		return err
+	}
+
+	channel := make(chan os.Signal, 1)
+	signal.Notify(channel, syscall.SIGWINCH)
+	// initial resize
+	channel <- syscall.SIGWINCH
+
+	// read output from container
+	go func() {
+		output := os.Stdout
+		_, _ = io.Copy(output, hijack.Conn)
+		channel <- syscall.SIGINT
+	}()
+
+	// send input to container
+	if c.Details.Config.OpenStdin {
+		go func() {
+			// the default escape key sequence: ctrl-p, ctrl-q
+			//
+			// shamelessly taken from docker/cli
+			escapeKeys := []byte{16, 17}
+			// Stop reading if escape sequence had been entered
+			input := term.NewEscapeProxy(os.Stdin, escapeKeys)
+			_, _ = io.Copy(hijack.Conn, input)
+			channel <- syscall.SIGINT
+		}()
+	}
+
+	for {
+		sig := <-channel
+		if sig == syscall.SIGWINCH {
+			err := c.Resize(fd)
+			if err != nil {
+				return err
+			}
+		} else {
+			err = term.RestoreTerminal(fd, oldState)
+			if err != nil {
+				return err
+			}
+
+			hijack.Close()
```

Comment:
this should be `defer hijack.Close()` right after we initialize it so that if we ever return early due to an error the session will be closed